### ISSUE TITLE
git verifyAuth failed on public Github repository

### ISSUE THREAD

**Issue by escalate** [semantic-release/semantic-release#699]
## Current behavior

During the CI run I receive a failure of the verifyAuth function (lib/git.js).
The git push in dry-run mode to my public Github repository does not work.
The call does not use the GH_TOKEN environemt variable to authenticate.
The failure seems to be not very crtitcal because the following steps are running without errors.
The step "Verify GitHub authentication" works correct so there is no real problem.

## Expected behavior

1. No simple verification test of the git repository because the connectivity to Github will be checked later.

OR

2. The verification test should use the GH_TOKEN environment variable to authenticate.

## Environment

- **semantic-release** version: latest
- CI environment: Travis CI https://github.com/escalate/loxone-ha-bridge-importer/blob/master/.travis.yml
- Plugins used: github, commit-analyzer, release-notes-generator
- **semantic-release** configuration: https://github.com/escalate/loxone-ha-bridge-importer/blob/master/.releaserc
- CI logs:

```
...
  py36: commands succeeded
  flake8: commands succeeded
  congratulations :)

travis_time:end:19ce6194:start=1520925358526575323,finish=1520925376365937399,duration=17839362076
[0K
[32;1mThe command "tox" exited with 0.[0m
travis_fold:start:after_success.1
[0Ktravis_time:start:079149b4
[0K$ nvm install lts/*
Downloading and installing node v8.10.0...
Downloading https://nodejs.org/dist/v8.10.0/node-v8.10.0-linux-x64.tar.xz...
Computing checksum with sha256sum
Checksums matched!
Now using node v8.10.0 (npm v5.6.0)

travis_time:end:079149b4:start=1520925376373571227,finish=1520925379187824204,duration=2814252977
[0Ktravis_fold:end:after_success.1
[0Ktravis_fold:start:after_success.2
[0Ktravis_time:start:02d84422
[0K$ npx travis-deploy-once "npx semantic-release --debug"
npx: installed 246 in 8.829s
[Travis Deploy Once]: There is only one job for this build.
npx: installed 387 in 13.296s
[Semantic release]: Running semantic-release version 15.0.3
2018-03-13T07:16:43.735Z semantic-release:git Error: Command failed: git push --dry-run https://github.com/escalate/loxone-ha-bridge-importer.git HEAD:master
remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/escalate/loxone-ha-bridge-importer.git/'


    at makeError (/home/travis/.npm/_npx/2975/lib/node_modules/semantic-release/node_modules/execa/index.js:169:9)
    at Promise.all.then.arr (/home/travis/.npm/_npx/2975/lib/node_modules/semantic-release/node_modules/execa/index.js:274:16)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
2018-03-13T07:16:43.741Z semantic-release:config options values: { branch: 'master',
  repositoryUrl: 'https://[secure]@github.com/escalate/loxone-ha-bridge-importer.git',
  tagFormat: 'v${version}',
  verifyConditions: [ '@semantic-release/github' ],
  analyzeCommits: [ '@semantic-release/commit-analyzer' ],
  verifyRelease: [],
  generateNotes: [ '@semantic-release/release-notes-generator' ],
  prepare: [],
  publish: [ '@semantic-release/github' ],
  success: [ '@semantic-release/github' ],
  fail: [ '@semantic-release/github' ],
  _: [],
  debug: true,
  '$0': '../../../.npm/_npx/2975/bin/semantic-release' }
[Semantic release]: Load plugin verifyConditions from @semantic-release/github
[Semantic release]: Load plugin analyzeCommits from @semantic-release/commit-analyzer
[Semantic release]: Load plugin generateNotes from @semantic-release/release-notes-generator
[Semantic release]: Load plugin publish from @semantic-release/github
[Semantic release]: Load plugin success from @semantic-release/github
[Semantic release]: Load plugin fail from @semantic-release/github
[Semantic release]: Run automated release from branch master
[Semantic release]: Call plugin verify-conditions
[Semantic release]: Verify GitHub authentication
2018-03-13T07:16:44.673Z semantic-release:get-last-release found tags: []
[Semantic release]: No git tag version found
[Semantic release]: No previous release found, retrieving all commits
[Semantic release]: Found 23 commits since last release
...
```

**Comment by pvdlg** [comment]
We check if the `repositoryUrl` configured in the via option or `package.json` allow to push to repo with transformation. Then if it doesn't we transform it and test again.
This has been implemented in #667 and implemented in #671.

What would be the point of removing this feature? In which way does it creates a problem?

**Comment by escalate** [comment]
The problem at the moment is that it creates ugly error messages.
In my mind errors should not be ignored as the above one.

I think it duplicates the functionality of the verifyConditions step.
If I use github or plain git the corresponding module should take care of the connectivity and not the main program.

I did not test it but if would add repositoryUrl like

**Comment by escalate** [comment]
I have to add more details to this bug report first.
The error message only occurs if I'm using the "--debug" toggle.
This is the output without the toggle.
```
  py36: commands succeeded
  flake8: commands succeeded
  congratulations :)

travis_time:end:00ebb0e9:start=1520927634655338421,finish=1520927659946117329,duration=25290778908
[0K
[32;1mThe command "tox" exited with 0.[0m
travis_fold:start:after_success.1
[0Ktravis_time:start:2cb31d24
[0K$ nvm install lts/*
Downloading and installing node v8.10.0...
Downloading https://nodejs.org/dist/v8.10.0/node-v8.10.0-linux-x64.tar.xz...
Computing checksum with sha256sum
Checksums matched!
Now using node v8.10.0 (npm v5.6.0)

travis_time:end:2cb31d24:start=1520927659954592288,finish=1520927662930050654,duration=2975458366
[0Ktravis_fold:end:after_success.1
[0Ktravis_fold:start:after_success.2
[0Ktravis_time:start:1c0934e5
[0K$ npx travis-deploy-once "npx semantic-release"
npx: installed 246 in 9.772s
[Travis Deploy Once]: There is only one job for this build.
npx: installed 387 in 12.623s
[Semantic release]: Running semantic-release version 15.0.3
[Semantic release]: Load plugin verifyConditions from @semantic-release/github
[Semantic release]: Load plugin analyzeCommits from @semantic-release/commit-analyzer
[Semantic release]: Load plugin generateNotes from @semantic-release/release-notes-generator
[Semantic release]: Load plugin publish from @semantic-release/github
[Semantic release]: Load plugin success from @semantic-release/github
[Semantic release]: Load plugin fail from @semantic-release/github
[Semantic release]: Run automated release from branch master
[Semantic release]: Call plugin verify-conditions
[Semantic release]: Verify GitHub authentication
[Semantic release]: Found git tag v1.0.0 associated with version 1.0.0
[Semantic release]: Found 1 commits since last release
[Semantic release]: Call plugin analyze-commits
[Semantic release]: Analyzing commit: ci: remove debug option for semantic-release
[Semantic release]: The commit should not trigger a release
[Semantic release]: Analysis of 1 commits complete: no release
[Semantic release]: There are no relevant changes, so no new version is released.

travis_time:end:1c0934e5:start=1520927662937964847,finish=1520927690524800785,duration=27586835938
[0Ktravis_fold:end:after_success.2
[0K
```

**Comment by pvdlg** [comment]
> The problem at the moment is that it creates ugly error messages.

The message appear only with `--debug`.

> In my mind errors should not be ignored as the above one.

We use the `git` to determine if we have push access. If we don't have push access the cli return an error (as a non `0` return code). In such case the error means that we don't have push access. So it's not an error in the sense that is it's an expected possible outcome.

> I think it duplicates the functionality of the verifyConditions step.
If I use github or plain git the corresponding module should take care of the connectivity and not the main program.

No because the core needs push access to create the Git tag. The core cannot rely on plugins to make this verification, as plugin are optional (that's the concept of plugins after all). So the core cannot rely on a plugins to verify something it requires. The core has to verify what it require, and the plugin verify what they require.

> I did not test it but if would add repositoryUrl like

It seems this sentence is incomplete.

**Comment by escalate** [comment]
Sorry this was my fault. I added my a comment draft too quick. I deleted the comment immediately.

In my previous comment I mentioned that the error only appears with debug mode enabled. 
That makes this bug very small for me.

I read the solution for issue #667. At this point I don't understand how the core should really work now.

> Use the unmodified configured repositoryUrl (without adding the credentials) if it works, otherwise do as of now and convert the URL to http/https and add the credentials. This allow users to push tags and commits via ssh while still using the GitHub/GitLab API.

The part "convert the URL to http/https and add the credentials" should not create the mentioned error.
Am I right?

For testing purpose I added the property repositoryUrl to .releaserc file. The error still occures in debug mode.

```
repositoryUrl: "https://github.com/escalate/loxone-ha-bridge-importer.git"
```

My expectation is that the core usees the GH_TOKEN if its present to test the push funtionality.
But it uses username and password which are not present.
Or is the usage of username and password only one part of the whole push tests?

Anyway the stack trace is not really nice. I think the smallest change could be to catch the error and print a nicer and more clearer message.

**Comment by pvdlg** [comment]
Once again, the process works as follow:
1. Check if the URL configured in `repositoryUrl`, or `package.json` or in the `origin` remote of the repo allows to push
2. If it does allow to push, use this one
3. If it doesn't allow to push, transform it to `https` or `http` and add the credentials
4. Test if get push access
5. If allows to push use it, if it doesn't throw an error

This is implemented this way as there is use cases in which users want to use the URL they configure without adding the credential. This is useful for example when they configure `GH_TOKEN` to access the API but want to use ssh keys for the git command line.

I really don't see the point of removing the ability to use a non transformed URL if it provides the proper push permission. We are not going to remove a feature that was requested, implemented and useful for some users, especially if having this feature doesn't create any issue and everything work as expected.

I understand the current implementation might not provide value for your specific use case, but that's not a reason to remove it, especially as it doesn't prevent your use case to work. Other users rely on that behavior.

**Comment by AVVS** [comment]
@pvdlg I might sound a bit frustrated, but how do people not run into issues with the following:

```
/**
 * Verify the write access authorization to remote repository with push dry-run.
 *
 * @param {String} origin The remote repository URL.
 * @param {String} branch The repositoru branch for which to verify write access.
 *
 * @return {Boolean} `true` is authorized to push, falsy otherwise.
 */
async function verifyAuth(origin, branch) {
  try {
    return (await execa('git', ['push', '--dry-run', origin, `HEAD:${branch}`])).code === 0;
  } catch (err) {
    debug(err);
  }
}
```

This simply asks the user to input "username" and "password" when pushing to https:// based URLs - and thats often the case for OSS projects.

I dont see any way to remove prompt for username/password from git when its not supplied, but at least set `timeout` option for execa to fail after 2-3 seconds or so - that should be enough to verify authentication via ssh

**Comment by pvdlg** [comment]
@AVVS this is not the same as what's reported in this issue, however it's a good point.

Git ask for user/password in the following conditions:
- The repo config `credential.helper` is set with a store that can ask for user/password
- The `GIT_TERMINAL_PROMPT` environment variable is not set to `0`
- The authentication fail

As far as I know most CI environment do not set `credential.helper` so the credential will not be asked when running on a CI. They would be asked when running locally, which is expected, so users can type them in the prompt.

I never encountered this issue, and we didn't had report that it happened to someone in the CI environment. Does it happened to you? Did that happen on the CI environment?

We could set a timeout for `execa` as you suggested and set `GIT_TERMINAL_PROMPT` to `0` to make sure the problem won't happens. However when running locally (for a dry-run or a local release) users would have to configure `GIT_CREDENTIALS`, `GH_TOKEN` or `GL_TOKEN` as they won't be prompted for user/password in case those are not set. It's not necessary a big issue as in most cases `GH_TOKEN` is required anyway, even for dry-runs.

If we make such fix, I'd like to have a way to reproduce the issue on a CI environment, so we can test we actually fix it.

**Comment by AVVS** [comment]
I run this on semaphore CI - you can test it there I suppose, they are free for OSS

**Comment by pvdlg** [comment]
Did you experience this issue on Semaphore? Do you have the semantic-release config and the CI logs when the problem happened?

Is what you describe is actually a problem you experienced or just a concern?

**Comment by AVVS** [comment]
I do experience the issue:


60:01
LINE WRAP OFF <https://semaphoreci.com/makeomatic/deploy/servers/git-ssh/deploys/28#>                                                                               
yarn run v1.3.2

                                                                               
$ semantic-release
[Semantic release]: Running semantic-release version 15.0.3
Username for 'https://github.com': execution expired

idk if it will show or not since its a deploy log - https://semaphoreci.com/makeomatic/deploy/servers/git-ssh/deploys/28 <https://semaphoreci.com/makeomatic/deploy/servers/git-ssh/deploys/28>
These are the settings it fails with - https://github.com/makeomatic/deploy/blob/master/.releaserc.json <https://github.com/makeomatic/deploy/blob/master/.releaserc.json> (tried to upgrade SR from 12.x to 15.x)

**Comment by pvdlg** [comment]
It seems I can't access the semaphore links you provided. I get a 404 page.

Can provide the semantic-release logs via a Gist or in this issue?

**Comment by AVVS** [comment]
I've actually included them earlier, since I suspected it would be the case:

```
$ semantic-release
[Semantic release]: Running semantic-release version 15.0.3
Username for 'https://github.com': execution expired
```

This is after the build timed out in 60 minutes :)

**Comment by AVVS** [comment]
I did debug the issue first before filing it here, it does happen during the aforementioned execa call (locally as well during the dry-run since I only have ssh credentials, and when its an https schema it would ask for username/password, which is probably not something I want to do anyway)

**Comment by pvdlg** [comment]
Ok I'm looking at it. For information if you want to use ssh rather than https you have to configure `repositoryUrl` as a git url (like `git@github.com:owner/repo.git`). This way the first call to `git --dry-run` will works if your ssh keys are properly set up, and semantic-release will use that URL.

**Comment by AVVS** [comment]
Yeah I understand, thats the case for the private repos and I have no problem with them, but for the public/OSS ones its preferable to set the repo URL to https

Thanks

**Comment by pvdlg** [comment]
After further investigations, Travis never prompt the user because it sets `GIT_ASKPASS=echo`.
Semaphore doesn't set this variable and prompt for the user/password.

On Semaphore, if `GIT_TERMINAL_PROMPT` is set to `0` there is no prompt and the git command errors out as expected.

It seems `GIT_TERMINAL_PROMPT` is a good solution but it was added in Git 2.3.0 The current version of Git is 2.16.1 but some CIs (CircleCI) use really old version of git.

So in order to be sure it won't happen on any CI I would propose to do the following when semantic-release runs on a CI:
- Set `GIT_TERMINAL_PROMPT` to `0`
- Set `GIT_ASKPASS` to `echo`
- Set a timeout when calling `execa`

**Comment by AVVS** [comment]
Seems reasonable to me

**Comment by pvdlg** [comment]
@AVVS the change has been released in 15.1.2.

@escalate I'm going to close this issue as it doesn't seems you are reporting an actual problem. If you experience a problem, please re-open this ticket or a new one.
